### PR TITLE
Rough support for sdformat robot descriptions

### PR DESCRIPTION
**Pull request by sloretz**
This is a quick hack to get rough sdformat support in joint_state_publisher.

@scpeters @iantheengineer FYI

Needed to support ros2/urdf#13

### CODE REVIEW

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 1]
```diff
@@ -161,7 +203,9 @@ def configure_robot(self, description):
         self.free_joints = {}
         self.joint_list = [] # for maintaining the original order of the joints
 
-        if robot.getElementsByTagName('COLLADA'):
+        if robot.getElementsByTagName('sdf'):
+            self.init_sdf(robot)
+        elif robot.getElementsByTagName('COLLADA'):
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I know you are just extending this, but this check seems too broad.  That is, if I understand https://docs.python.org/3.6/library/xml.dom.html#xml.dom.Document.getElementsByTagName correctly, this will look in the entire XML tree for tags of this name.  I think this has some (limited) potential to produce false positives.  If someone was using COLLADA, but named one of their joints 'sdf', I think we would attempt to parse it as SDF.  Is there something else we can check in the document that is more specific?  The root element name, for instance?

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 2]
```diff
@@ -59,6 +59,46 @@ def _init_joint(self, minval, maxval, zeroval):
...
 
+    def init_sdf(self, robot):
+        robot = robot.getElementsByTagName('model')[0]
+        # Find all non-fixed joints
+        for child in robot.childNodes:
+            if child.nodeType is child.TEXT_NODE:
+                continue
+            if child.localName == 'joint':
+                jtype = child.getAttribute('type')
+                if jtype in ['gearbox', 'revolute2', 'ball', 'screw', 'universal', 'fixed']:
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Uber-nit: a tuple is *slightly* faster for a fixed list.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 3]
```diff
@@ -59,6 +59,46 @@ def _init_joint(self, minval, maxval, zeroval):
             joint['effort'] = 0.0
         return joint
 
+    def init_sdf(self, robot):
+        robot = robot.getElementsByTagName('model')[0]
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py:63]: How about SDF files that contain many models?  How should we handle those?

**Review by clalancette — COMMENTED** [review]
I just took a quick look here, pointed out a couple of potential issues I saw.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 3 above)
Comment (sloretz) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py:63]: I think we should error, and only allow one model

Comment (sloretz) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py:63]: I think it should error, and only allow documents with one model.

https://github.com/ros/sdformat_urdf/pull/1/files#diff-12b109e5d429371a684ac656c3aa2037R59-R64

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 1 above)
Comment (sloretz) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Yeah, it seems too broad to me too. I think checking the root element would be a better solution. The C++ parser is handling it a little different by scoring based on the location of the first tag text, like `<COLLADA`

https://github.com/ros/collada_urdf/pull/39/files#diff-f13e149a001d68baae11ac5da7f08dc5R56-R59

but that would be fooled by the text being in a comment in the xml.

```xml
<!-- this is not a <COLLADA> file -->
<sdf>
...
```

It would be nice if joint_state_publisher used the C++ parser so this would only have to be fixed in one place.

Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: > It would be nice if joint_state_publisher used the C++ parser so this would only have to be fixed in one place.

While I somewhat agree, I also don't relish the thought of putting a CPython binding on top of the URDF ecosystem.  At the very least, it means we don't have a universal wheel for Python anymore.

Comment (sloretz) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: > I also don't relish the thought of putting a CPython binding on top of the URDF ecosystem. At the very least, it means we don't have a universal wheel for Python anymore.

Where were you thinking about the CPython binding living? `urdf_parser_py`? I was thinking of `joint_state_publisher` having a CPython extension that calls the urdf C++ api to parse the robot description and return an object with the joint info joint_state_publisher needs.

Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: > Where were you thinking about the CPython binding living? `urdf_parser_py`? I was thinking of `joint_state_publisher` having a CPython extension that calls the urdf C++ api to parse the robot description and return an object with the joint info joint_state_publisher needs.

I didn't have a specific place in mind.

But even in the case you are suggesting, its not clear to me that providing a CPython extension for joint_state_publisher is a win.  It seems to me the pros and cons of switching to a CPython extension:

Pros:
* Reduce duplication in parsing the XML.
* Get support for new formats for "free" once they are in urdf_parser

Cons:
* Replace Python code most Python programmers can understand with more complex CPython binding.
* joint_state_publisher is no longer a [Pure python](https://packaging.python.org/guides/distributing-packages-using-setuptools/#pure-python-wheels) wheel.

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2020-07-16/15468/1

**Comment by clalancette** [comment]
@sloretz Just out of curiousity, is this something that you are still interested in moving forward with?

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 3 above)
Comment (sloretz) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py:63]: SDFormat no longer allows files with multiple root-level models, so I'll resolve this thread.

https://github.com/gazebosim/sdformat/pull/619

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 1 above)
Comment (sloretz) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: It's been a few years. Looking back, reusing the existing URDF/SDformat parsers instead of partially implementing parsers in this package seems like a big advantage, and using another language we're already extensively using in ROS seems like a small disadvantage.

I'm not about the pure python wheel con. It looks like [`colcon` doesn't build packages as wheels](https://github.com/colcon/colcon-core/blob/99c03262ad7edd4af5954c424c9248e0c4fe7307/colcon_core/task/python/build.py#L102-L104), and if someone were building the package without using colcon I'm not sure what they would find more advantageous.

File: joint_state_publisher/test/multi_joint_robot.sdf [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+          <xyz>0 0 1</xyz>
+          <limit>
+            <lower>-1</lower>
+            <upper>1</upper>
+          </limit>
+        </axis>
+    </joint>
+
+  </model>
+</sdf>
```
Comment (clalancette) [inline joint_state_publisher/test/multi_joint_robot.sdf]: Nit: add a trailing newline at the end of the file.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 5]
```diff
@@ -161,10 +204,16 @@ def configure_robot(self, description):
...
 
-        if robot.getElementsByTagName('COLLADA'):
+        # Get root tag to parse file format
+        root = robot.documentElement
+        if root.tagName == 'sdf':
+            self.init_sdf(robot)
+        elif root.tagName == 'COLLADA':
             self.init_collada(robot)
-        else:
+        elif root.tagName == 'urdf':
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I don't think that this is correct; as far as I know, `urdf` files start with the `robot` tag: https://github.com/ros/urdfdom/blob/b5fffec979cabb9563f6f0ad3aebd92bdc978bae/xsd/urdf.xsd#L300

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 6]
```diff
@@ -161,10 +204,16 @@ def configure_robot(self, description):
...
             self.init_urdf(robot)
+        else:
+            self.get_logger().error('Cannot parse file of type %s' % root.tagName)
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: If we are going to error here, we should also `return`.

Comment (clalancette) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Also, I can't put a comment on the code below due to GitHub UI limitations, but in `__init__` we should change the name of the parameter from `urdf_file` to just `description_file` or something like that.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 7]
```diff
@@ -73,9 +114,11 @@ def init_collada(self, robot):
...
 
                 if joint:
-                    limit = joint.getElementsByTagName('limits')[0]
-                    minval = float(limit.getElementsByTagName('min')[0].childNodes[0].nodeValue)
-                    maxval = float(limit.getElementsByTagName('max')[0].childNodes[0].nodeValue)
+                    limit_list = joint.getElementsByTagName('limits')
+                    if limit_list:
+                        limit = limit_list[0]
+                        minval = float(limit.getElementsByTagName('min')[0].childNodes[0].nodeValue)
+                        maxval = float(limit.getElementsByTagName('max')[0].childNodes[0].nodeValue)
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I'm not convinced we should do this particular fix here.  That is, this is a good fix to have, and I think we should make the fix, but I also think that if we do this we should do it consistently for all of `init_urdf`, `init_collada`, and `init_sdf`.  So I'll suggest we just do this in a separate PR.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 8]
```diff
@@ -59,6 +59,47 @@ def _init_joint(self, minval, maxval, zeroval):
...
+                if jtype == 'continuous':
+                    minval = -math.pi
+                    maxval = math.pi
+                else:
+                    try:
+                        limit = child.getElementsByTagName('limit')[0]
+                        minval = float(limit.getElementsByTagName('lower')[0].firstChild.data)
+                        maxval = float(limit.getElementsByTagName('upper')[0].firstChild.data)
+                    except ValueError:
+                        self.get_logger().warn('%s limits are not valid!' % name)
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I think this is missing a `continue`:

```suggestion
                        self.get_logger().warn('%s limits are not valid!' % name)
                        continue
```

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 9]
```diff
@@ -59,6 +59,47 @@ def _init_joint(self, minval, maxval, zeroval):
             joint['effort'] = 0.0
         return joint
 
+    def init_sdf(self, robot):
+        robot = robot.getElementsByTagName('model')[0]
+        # Find all non-fixed joints
+        for child in robot.childNodes:
+            if child.nodeType is child.TEXT_NODE:
+                continue
+            if child.localName == 'joint':
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I know the rest of the code isn't like this, but I think this would be easier to read if we got out early.  That is, change this line to be:

```
if child.localName != 'joint':
    continue
```

and then unindent the rest of the code one level (we should do the same for `init_urdf` and `init_collada`, but that should be a separate PR).

**Review by clalancette — CHANGES_REQUESTED** [review]

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 5 above)
Comment (quarkytale) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: Should I go for checking both urdf and robot tags or no checks directly parsing it as urdf?

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 7 above)
Comment (quarkytale) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: https://github.com/ros/joint_state_publisher/pull/94

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 9 above)
Comment (quarkytale) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: https://github.com/ros/joint_state_publisher/pull/94

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 10]
```diff
@@ -406,13 +458,13 @@ def main():
     # Strip off the ROS 2-specific command-line arguments
     stripped_args = rclpy.utilities.remove_ros_args(args=sys.argv)
     parser = argparse.ArgumentParser()
-    parser.add_argument('urdf_file', help='URDF file to use', nargs='?', default=None)
+    parser.add_argument('description_file', help='Robot description file to use', nargs='?', default=None)
 
     # Parse the remaining arguments, noting that the passed-in args must *not*
     # contain the name of the program.
     parsed_args = parser.parse_args(args=stripped_args[1:])
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py:462]: We can't do this exactly like this, because we would break backwards compatibility with anyone currently using the `urdf_file` argument.

Instead, I think we should deprecate the `urdf_file` argument, but still honor it for now.  Something like this (untested):

```suggestion
    parser.add_argument('urdf_file', help='URDF file to use (deprecated, use "description_file" instead)', nargs='?', default=None)
    parser.add_argument('description_file', help='Robot description file to use', nargs='?', default=None)

    # Parse the remaining arguments, noting that the passed-in args must *not*
    # contain the name of the program.
    parsed_args = parser.parse_args(args=stripped_args[1:])
    
    if parsed_args.urdf_file and parsed_args.description_file:
        raise Exception('Only one of "urdf_file" or "description_file" can be specified')
    elif parsed_args.urdf_file:
        warnings.warn('The "urdf_file" argument is deprecated, switch to using "description_file"')
        parsed_args.description_file = parsed_args.urdf_file
```

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 11]
```diff
@@ -59,6 +59,51 @@ def _init_joint(self, minval, maxval, zeroval):
...
+        for child in robot.childNodes:
+            if child.nodeType is child.TEXT_NODE:
+                continue
+            if child.localName != 'joint':
+                continue
+            jtype = child.getAttribute('type')
+            if jtype in ('gearbox', 'revolute2', 'ball', 'screw', 'universal', 'fixed'):
+                continue
+            name = child.getAttribute('name')
+            self.joint_list.append(name)
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I think this line should be moved below, right before we do `self.free_joints[name] = joint`.

The main reason is that there are additional checks down below, so if a joint violates one of the limits, then we could get into a state where it was added to `self.joint_list`, but not `self.free_joints`.  To avoid that, we shouldn't change any object state until after we have done all of our checks.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 12]
```diff
@@ -161,10 +206,17 @@ def configure_robot(self, description):
...
+        if root.tagName == 'sdf':
+            self.init_sdf(robot)
+        elif root.tagName == 'COLLADA':
             self.init_collada(robot)
-        else:
+        elif root.tagName == 'urdf' or root.tagName == 'robot':
             self.init_urdf(robot)
+        else:
+            self.get_logger().error('Cannot parse file of type %s' % root.tagName)
+            return
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: After thinking about it more, for maximum compatibility we should probably just leave this as `else`.  It is somewhat incorrect in that we'll try to parse *anything* else as URDF, but it also keeps us the most compatible with what is already there.

```suggestion
        else:
            self.init_urdf(robot)
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left a few more things to fix, but this is a lot closer to being ready.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 10 above)
Comment (quarkytale) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py:462]: My bad, fixed it in https://github.com/ros/joint_state_publisher/pull/95, still running some tests

**Comment by clalancette** [comment]
@quarkytale Now that I've merged in #94, this one has a merge conflict.  If you could resolve that, I'll then do another review here.  Thanks.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 13]
```diff
@@ -408,13 +457,15 @@ def main():
     # Strip off the ROS 2-specific command-line arguments
     stripped_args = rclpy.utilities.remove_ros_args(args=sys.argv)
     parser = argparse.ArgumentParser()
-    parser.add_argument('urdf_file', help='URDF file to use', nargs='?', default=None)
+    parser.add_argument('urdf_file', help='URDF file to use (deprecated, use "description_file" instead)', nargs='?', default=None)
+    parser.add_argument('description_file', help='Robot description file to use', nargs='?', default=None)
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: I'm sorry, I made a huge mistake here.

This isn't a named option, it is just the (optional) option you pass on the command-line.  That is, it works like:

`ros2 run joint_state_publisher joint_state_publisher /path/to/urdf`

So in that case, we *can* just rename it, since the name just ends up being a local variable.  I think we should change this to:

```suggestion
    parser.add_argument('description_file', help='Robot description file to use', nargs='?', default=None)
```

Because that will just be simpler to understand.  Then we can also remove the `if not parsed_args.description_file` check below.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py [hunk 14]
```diff
@@ -59,6 +59,51 @@ def _init_joint(self, minval, maxval, zeroval):
...
+            if jtype == 'continuous':
+                minval = -math.pi
+                maxval = math.pi
+            else:
+                try:
+                    limit_list = child.getElementsByTagName('limit')
+                    if limit_list:
+                        limit = limit_list[0]
+                        minval = float(limit.getElementsByTagName('lower')[0].firstChild.data)
+                        maxval = float(limit.getElementsByTagName('upper')[0].firstChild.data)
```
Comment (clalancette) [inline joint_state_publisher/joint_state_publisher/joint_state_publisher.py]:
```suggestion
                    limit = child.getElementsByTagName('limit')[0]
                    minval = float(limit.getElementsByTagName('lower')[0].firstChild.data)
                    maxval = float(limit.getElementsByTagName('upper')[0].firstChild.data)
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left two more things to fix, then I think we can get this in.

File: joint_state_publisher/joint_state_publisher/joint_state_publisher.py (hunk 13 above)
Comment (quarkytale) [reply joint_state_publisher/joint_state_publisher/joint_state_publisher.py]: That's what I was wondering when I didn't find any mention of it in the test cases. Reverting back

**Review by clalancette — APPROVED** [review]
